roles/mongodb/defaults/main.yml

```diff
@@ -2,3 +2,25 @@
 # MongoDB version to install
 # Should be '3.2' or '3.4'
 mongodb_version: "3.4"
+
+# do not enable mongo auth by default. Make sure to add required users in mongodb_users if auth is enabled.
+mongodb_enable_auth: no
+
+# the directory used to store generated credentials in
+mongodb_creds_dir: '.'
+
+# when adding auth, the login credentials to use
+mongodb_host: 127.0.0.1
+mongodb_port: 27017
+mongodb_admin_username: admin
+mongodb_admin_password: "{{ lookup('password', '{{ mongodb_creds_dir }}/mongodb-' + inventory_hostname + '-' + mongodb_username + ' length=42' ) }}"
+
+# Additional users to add.
+mongodb_users: []
+#  - db: st2
+#    username: st2mongo
+#    password: "{{ lookup('password', '{{ mongodb_creds_dir }}/mongodb-' + inventory_hostname + '-' + st2mongo_username + ' length=42' ) }}"
+#    roles: readWrite
+
+# whether or not to force a password update for any users in mongodb_users
+mongodb_force_update_password: no
```

Comment:
Any reasons why do we need to set `update_password` to `no`?

I guess we'll see unexpected behavior while changing the password in the following example:

```yaml
          vars:
            mongodb_users:
              - db: st2
                username: st2mongo
                password: "new-password"
```

I think it's always desired to have password updated when specifying new one for the mongo user.

Reply:
We need to make sure `update_password: always` is idempotent. I think I was having issues with that so I switched to `update_password: no`. Re-reading the docs it looks like `always` is supposed to be idempotent, so maybe I'm remembering wrong.

I'm okay switching this to always, or to an update_password var that defaults to always.

Reply:
https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/database/mongodb/mongodb_user.py#L440-L455
mongodb_user is not idempotent. If you do update_password: always then every time that task runs it will report "changed" even if the password is the same as what is already there.
If you do update_password: on create, then the module checks to see if the user (a) exists on the db, and (b) has the same roles, and then it only adds the user if it's not there or the roles have changed.
So, we can do update_password: always only if we're ok with having continuously report changed on every run.

Reply:
There is a comment in the module about checking the password:
```python
            # Here we can  check password change if mongo provide a query for that : https://jira.mongodb.org/browse/SERVER-22848
            #newuinfo = user_find(client, user, db_name)
            #if uinfo['role'] == newuinfo['role'] and CheckPasswordHere:
            #    module.exit_json(changed=False, user=user)
```
In that [mongo Jira issue](https://jira.mongodb.org/browse/SERVER-22848), they prefer:
> all password checks should always be audited as login attempts for security purposes

So, we to try and make this more idempotent we can try and login with the username/password to see if a password update is required. That will have the side effect of having a login attempt logged every time we run the playbook. If that's ok, great. We can either do that login attempt by

1. Change the module to make it do the test login.
2. Add some set of tasks to the playbook that try to login before adding the user... I don't know what that would look like. Maybe we can use the [mongodb lookup](https://docs.ansible.com/ansible/devel/plugins/lookup/mongodb.html) to login and read something (what? How does login failure via lookup act?) Then, if that login was successful, use `update_password: on create` or if unsuccessful (user doesn't exist or password failed) then `update_password: always`.

Reply:
Thanks for the research, that makes sense :+1: 
Would be good to include a code comment for future contributors about the idempotence failure in Ansible module, so we won't try to change this again.